File: Homework/Hang_Nguyen/lesson_1/happy-number.py
class Solution:
    def squareSumDigit(self, n):
        s = 0
        while n:
            s += (n%10) * (n%10)
            n = int(n/10)
        return s

    def isHappy(self, n):
        slow = n
        fast = n
        while True:
            slow = self.squareSumDigit(slow)
            fast = self.squareSumDigit(self.squareSumDigit(fast))
            if slow == fast: break
        return slow == 1

# TEST CASE
solution = Solution()

print("TEST 1 is " + str(solution.isHappy(19)))
print("TEST 2 is " + str(solution.isHappy(2)))
print("TEST 3 is " + str(solution.isHappy(3)))
print("TEST 4 is " + str(solution.isHappy(7)))

File: Homework/Truong_Tran/lesson_1/HappyNumber.java
package hackerank;

public class HappyNumber {

    public static boolean isHappy(int n) {
        int p1 = sumSquare(n);  // slow pointer
        int p2 = sumSquare(sumSquare(n));  // fast pointer
        while (p1 != p2) {
            p1 = sumSquare(p1);
            p2 = sumSquare(sumSquare(p2));
        }
        if (p1 == 1) {
            return true;
        } else {
            return false;
        }
    }

    public static int sumSquare(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n = n/10;
        }
        return sum;
    }
}
